src/peft/tuners/lora/tp_layer.py

```diff
@@ -0,0 +1,160 @@
+        self.backend = backend
+        self.is_paralle_a = is_paralle_a
+
+    def update_layer(self, adapter_name, r, lora_alpha, lora_dropout, init_lora_weights, **parallel_linear_kwargs):
+        if r <= 0:
+            raise ValueError(f"`r` should be a positive integer value but the value passed is {r}")
+        self.r[adapter_name] = r
+        self.lora_alpha[adapter_name] = lora_alpha
+        if lora_dropout > 0.0:
+            lora_dropout_layer = nn.Dropout(p=lora_dropout)
+        else:
+            lora_dropout_layer = nn.Identity()
+
+        self.lora_dropout.update(nn.ModuleDict({adapter_name: lora_dropout_layer}))
+
+        config = parallel_linear_kwargs["config"]
+        # lora needs to be forced to upgrade to 32-bit precision, otherwise it will overflow
+        config.params_dtype = torch.float32
+        init_method = parallel_linear_kwargs.get("init_method", init.xavier_normal_)
+        input_is_parallel = parallel_linear_kwargs.get("input_is_parallel", True)
+        gather_output = parallel_linear_kwargs.get("gather_output", False)
+        if r > 0:
+            if self.is_paralle_a:
+                lora_a = self.backend.RowParallelLinear(
+                    input_size=self.in_features,
+                    output_size=r,
+                    bias=False,
+                    input_is_parallel=input_is_parallel,
+                    skip_bias_add=True,
+                    init_method=init_method,
+                    config=config,
+                )
+                lora_b = nn.Linear(in_features=r, out_features=self.out_features, bias=False, dtype=torch.float32)
+            else:
+                lora_a = nn.Linear(in_features=self.in_features, out_features=r, bias=False, dtype=torch.float32)
+                lora_b = self.backend.ColumnParallelLinear(
+                    input_size=r,
+                    output_size=self.out_features,
+                    bias=False,
+                    gather_output=gather_output,
+                    init_method=init_method,
+                    config=config,
+                )
+            self.lora_A[adapter_name] = lora_a
+            self.lora_B[adapter_name] = lora_b
+            self.scaling[adapter_name] = lora_alpha / r
+        if init_lora_weights:
+            self.reset_lora_parameters(adapter_name)
+
+        weight = getattr(self.get_base_layer(), "weight", None)
+        if weight is not None:
+            # the layer is already completely initialized, this is an update
+            if weight.dtype.is_floating_point or weight.dtype.is_complex:
+                self.to(weight.device, dtype=weight.dtype)
+            else:
+                self.to(weight.device)
+        self.set_adapter(self.active_adapters)
+
+
+def get_lora_paralle_linear(
+    base_layer, adapter_name: str, backend, r: int = 0, lora_alpha: int = 1, lora_dropout: float = 0.0, **kwargs
+):
+    class LoraParallelLinear(nn.Module, LoraParalleLayer):
```

Comment:
Why is this class defined inside the function, wouldn't it work if we define it on root level? We could pass the `backend` argument as an `__init__` parameter.

If this is changed, I think we can also remove `get_lora_paralle_linear` completely. Furthermore, I think we could remove the `LoraParalleLayer` class completely, which doesn't do much at the moment, only introducing the `update_layer` method. If we later want to add support for more layer types besides `Linear`, we can still factor this out into a mixin class for instance.

Reply:
> Why is this class defined inside the function, wouldn't it work if we define it on root level? We could pass the `backend` argument as an `__init__` parameter.
> 
> If this is changed, I think we can also remove `get_lora_paralle_linear` completely. Furthermore, I think we could remove the `LoraParalleLayer` class completely, which doesn't do much at the moment, only introducing the `update_layer` method. If we later want to add support for more layer types besides `Linear`, we can still factor this out into a mixin class for instance.

Yes, I don't need `get_lora_paralle_linear` and `LoraParalleLayer` this base class~